File: src/utils/urlMetadata/threadsParser.js
const cheerio = require('cheerio');
const logger = require('../logger');

module.exports = {
    match: (url) => {
        try {
            const u = new URL(url);
            return u.hostname === 'www.threads.net' || u.hostname === 'threads.net' || u.hostname === 'www.threads.com' || u.hostname === 'threads.com';
        } catch (e) {
            return false;
        }
    },
    parse: async (url) => {
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), 10000);

            // Fetch directly from Threads using native node fetch
            const response = await fetch(url, {
                signal: controller.signal
            });
            clearTimeout(timeoutId);

            if (!response.ok) return null;

            const html = await response.text();
            const $ = cheerio.load(html);

            const title = $('meta[property="og:title"]').attr('content') || '';
            let description = $('meta[property="og:description"]').attr('content') || '';
            const imageUrl = $('meta[property="og:image"]').attr('content');

            if (!title) return null;

            if (description.length > 200) {
                description = description.substring(0, 200) + '...';
            }

            const resultText = [
                `[Threads 推文 | ${title}]`,
                `${description}`
            ].join('\n');

            const images = [];
            if (imageUrl) {
                images.push(imageUrl);
            }

            return {
                text: resultText,
                images: images
            };
        } catch (err) {
            logger.warn(`Threads 解析失敗 (${url}): ${err.message}`);
            return null;
        }
    }
};

File: src/utils/urlMetadata/threadsParser.py
import urllib.parse
import requests
from bs4 import BeautifulSoup
from ..logger import logger

def match(url_string):
    try:
        parsed = urllib.parse.urlparse(url_string)
        hostname = parsed.hostname.replace('www.', '') if parsed.hostname else ''
        return hostname in ['threads.net', 'threads.com']
    except Exception:
        return False

def parse(url_string):
    try:
        response = requests.get(
            url_string,
            timeout=10
        )
        
        if not response.ok:
            return None
            
        soup = BeautifulSoup(response.text, 'html.parser')
        og_title = soup.find('meta', property='og:title')
        title = og_title['content'] if og_title else ''
        
        if not title:
            return None
            
        og_desc = soup.find('meta', property='og:description')
        description = og_desc['content'] if og_desc else ''
        
        og_img = soup.find('meta', property='og:image')
        image_url = og_img['content'] if og_img else None
        
        if len(description) > 200:
            description = description[:200] + '...'
            
        result_text = f"[Threads 推文 | {title}]\n{description}"
        
        images = []
        if image_url:
            images.append(image_url)
            
        return {
            'text': result_text,
            'images': images
        }
        
    except Exception as e:
        logger.warning(f"Threads 解析失敗 ({url_string}): {str(e)}")
        return None
